container-search/src/main/java/com/yahoo/search/dispatch/SearchPath.java

```diff
@@ -0,0 +1,253 @@
+     * @throws InvalidSearchPathException
+     *             if the searchPath is malformed
+     * @return list of nodes chosen with the search path, or an empty list in which
+     *         case some other node selection logic should be used
+     */
+    public static List<SearchCluster.Node> selectNodes(String searchPath, SearchCluster cluster) {
+        Optional<SearchPath> sp = SearchPath.fromString(searchPath);
+        if (sp.isPresent()) {
+            return sp.get().mapToNodes(cluster);
+        } else {
+            return Collections.emptyList();
+        }
+    }
+
+    public static Optional<SearchPath> fromString(String path) {
+        if (path == null || path.isEmpty()) {
+            return Optional.empty();
+        }
+        if (path.indexOf(';') >= 0) {
+            return Optional.empty(); // multi-level not supported at this time
+        }
+        try {
+            SearchPath sp = parseElement(path);
+            if (sp.isEmpty()) {
+                return Optional.empty();
+            } else {
+                return Optional.of(sp);
+            }
+        } catch (NumberFormatException | InvalidSearchPathException e) {
+            throw new InvalidSearchPathException("Invalid search path: " + path);
+        }
+    }
+
+    private final List<Part> parts;
+    private final Integer row;
+
+    private SearchPath(List<Part> parts, Integer row) {
+        this.parts = parts;
+        this.row = row;
+    }
+
+    private List<SearchCluster.Node> mapToNodes(SearchCluster cluster) {
+        if (cluster.groups().isEmpty()) {
+            return Collections.emptyList();
+        }
+
+        SearchCluster.Group group = selectGroup(cluster);
+
+        if (parts.isEmpty()) {
+            return group.nodes();
+        }
+        Set<Integer> wanted = new HashSet<>();
+        int max = group.nodes().size();
+        for (Part part : parts) {
+            wanted.addAll(part.matches(max));
+        }
+        // ordering by distribution key might not be equal to ordering in services.xml
+        List<SearchCluster.Node> sortedByDistKey = new ArrayList<>(group.nodes());
+        sortedByDistKey.sort(Comparator.comparingInt(SearchCluster.Node::key));
+
+        List<SearchCluster.Node> ret = new ArrayList<>();
+        for (int idx : wanted) {
+            ret.add(sortedByDistKey.get(idx));
+        }
+        return ret;
+    }
+
+    private boolean isEmpty() {
+        return parts.isEmpty() && row == null;
+    }
+
+    private Group selectGroup(SearchCluster cluster) {
+        // ordering by group id might not be equal to ordering in services.xml
+        ImmutableMap<Integer, SearchCluster.Group> byId = cluster.groups();
+        List<Integer> sortedKeys = new ArrayList<>(byId.keySet());
+        Collections.sort(sortedKeys);
+
+        if (row != null && row < sortedKeys.size()) {
+            return byId.get(sortedKeys.get(row));
+        }
+
+        // pick "anything": try to find the first working
```

Comment:
Shouldn't we throw an IllegalArgumentException here?

Reply:
The Search API doc doesn't mention overflows and the system test only tests against node overflows (which are just dropped). I'll switch over to an error since it's probably the path of least surprise.